### PR TITLE
typing fixes

### PR DESCRIPTION
**Pull request by Illviljan**
Newer version of mypy has found new issues with the typing.

### CODE REVIEW

**Comment by Illviljan** [comment]
```
Found 15 errors in 2 files (checked 11 source files)
flox/core.py:219: error: Argument 1 to "tuple" has incompatible type "Union[ExtensionDtype, str, dtype[generic], Type[object]]"; expected "Iterable[str]"  [arg-type]
flox/core.py:508: error: Incompatible types in assignment (expression has type "Tuple[None, ...]", variable has type "Optional[Tuple[Index, ...]]")  [assignment]
flox/core.py:512: error: Argument 2 to "zip" has incompatible type "Optional[Tuple[Index, ...]]"; expected "Iterable[Index]"  [arg-type]
flox/core.py:545: error: Argument 1 to "append" of "list" has incompatible type "IntervalIndex[Any]"; expected "ndarray[Any, dtype[Any]]"  [arg-type]
flox/core.py:559: error: Incompatible types in assignment (expression has type "ndarray[Any, Any]", variable has type "Union[Index, str, bytes, date, timedelta, datetime64, timedelta64, int, float, integer[Any], floating[Any], complex]")  [assignment]
flox/core.py:559: error: Argument 1 to "factorize" has incompatible type "Union[ndarray[Any, Any], Any]"; expected "Sequence[Any]"  [arg-type]
flox/core.py:750: error: Argument "expected_groups" to "factorize_" has incompatible type "Tuple[Optional[Index]]"; expected "Optional[Tuple[Index, ...]]"  [arg-type]
flox/core.py:1581: error: Argument 1 to "append" of "list" has incompatible type "Union[Sequence[Any], ndarray[Any, Any], Index, None]"; expected "Optional[Index]"  [arg-type]
flox/core.py:1584: error: Argument 1 to "from_breaks" of "IntervalIndex" has incompatible type "Union[Sequence[Any], ndarray[Any, Any], Index]"; expected "Union[Sequence[int], ndarray[Any, dtype[signedinteger[_64Bit]]], ndarray[Any, dtype[signedinteger[_32Bit]]], ndarray[Any, dtype[signedinteger[Any]]], Series[int], _IntIndexType]"  [arg-type]
flox/core.py:1643: error: Argument "expected_groups" to "factorize_" has incompatible type "Tuple[Optional[Index], ...]"; expected "Optional[Tuple[Index, ...]]"  [arg-type]
flox/xarray.py:333: error: Incompatible types in assignment (expression has type "Union[Sequence[Any], ndarray[Any, Any], Index]", variable has type "Index")  [assignment]
flox/xarray.py:449: error: Item "Sequence[Any]" of "Union[Sequence[Any], ndarray[Any, Any], Index, None]" has no attribute "values"  [union-attr]
flox/xarray.py:449: error: Item "ndarray[Any, Any]" of "Union[Sequence[Any], ndarray[Any, Any], Index, None]" has no attribute "values"  [union-attr]
flox/xarray.py:449: error: Item "None" of "Union[Sequence[Any], ndarray[Any, Any], Index, None]" has no attribute "values"  [union-attr]
Error: Process completed with exit code 1.
```

File: flox/core.py [hunk 1]
```diff
@@ -77,7 +81,7 @@
 # This dummy axis is inserted using np.expand_dims
 # and then reduced over during the combine stage by
 # _simple_combine.
-DUMMY_AXIS = -2
+DUMMY_AXIS = -2 + 0
```
Comment (Illviljan) [inline flox/core.py:84]:
```suggestion
DUMMY_AXIS = -2
```

File: flox/core.py [hunk 2]
```diff
@@ -219,8 +223,14 @@ def find_group_cohorts(labels, chunks, merge: bool = True) -> dict:
...
     # these are chunks where a label is present
     label_chunks = pd.Series(which_chunk).groupby(raveled).unique()
+
     # These invert the label_chunks mapping so we know which labels occur together.
-    chunks_cohorts = tlz.groupby(lambda x: tuple(label_chunks.get(x)), label_chunks.keys())
+    def invert(x) -> tuple[np.ndarray, ...]:
+        arr = label_chunks.get(x)
+        return tuple(arr)  # type: ignore [arg-type] # pandas issue?
+
+    # chunks_cohorts = tlz.groupby(lambda x: (label_chunks.get(x),), label_chunks.keys())
```
Comment (Illviljan) [inline flox/core.py:232]:
```suggestion
```

File: flox/core.py [hunk 3]
```diff
@@ -1584,18 +1596,33 @@ def _assert_by_is_aligned(shape: tuple[int, ...], by: T_Bys):
...
+            e = ex
             if sort:
-                ex = ex.sort_values()
-            out.append(ex)
+                e = e.sort_values()
+            out.append(e)
         elif ex is not None:
             if isbin_:
-                out.append(pd.IntervalIndex.from_breaks(ex))
+                out.append(pd.IntervalIndex.from_breaks(ex))  # type: ignore [arg-type] # TODO: what do we want here?
```
Comment (Illviljan) [inline flox/core.py:1625]: This one is surprisingly difficult to figure out and now I give up.

**Comment by Illviljan** [comment]
Ready for review @dcherian . I gave up a little and added a few ignores.  I think the errors are unrelated, #245.

**Comment by Illviljan** [comment]
@dcherian , finally all green! :)

File: flox/core.py [hunk 4]
```diff
@@ -98,7 +102,7 @@ def _is_first_last_reduction(func: T_Agg) -> bool:
     return isinstance(func, str) and func in ["nanfirst", "nanlast", "first", "last"]
 
 
-def _get_expected_groups(by: T_By, sort: bool) -> pd.Index:
+def _get_expected_groups(by: T_By, sort: bool) -> T_ExpectIndex:
```
Comment (dcherian) [inline flox/core.py:105]: Please leave this as `pd.Index`. https://matthewrocklin.com/avoid-indirection.html

File: flox/core.py [hunk 5]
```diff
@@ -853,7 +862,8 @@ def _finalize_results(
     """
     squeezed = _squeeze_results(results, axis)
 
-    if agg.min_count > 0:
+    min_count = agg.min_count if agg.min_count is not None else 0
```
Comment (dcherian) [inline flox/core.py:865]: I'd like to avoid `agg.min_count` being non-integer. There must be a bug somewhere?

File: flox/core.py [hunk 6]
```diff
@@ -1584,18 +1596,33 @@ def _assert_by_is_aligned(shape: tuple[int, ...], by: T_Bys):
...
+) -> T_ExpectIndexOptTuple:
+    out: list[T_ExpectIndexOpt] = []
     for ex, isbin_ in zip(expected_groups, isbin):
         if isinstance(ex, pd.IntervalIndex) or (isinstance(ex, pd.Index) and not isbin_):
+            e = ex
             if sort:
-                ex = ex.sort_values()
-            out.append(ex)
+                e = e.sort_values()
+            out.append(e)
```
Comment (dcherian) [inline flox/core.py:1622]: Can we revert this please?

File: flox/xarray.py [hunk 7]
```diff
@@ -326,21 +328,23 @@ def xarray_reduce(
...
                     f"Please provided bin edges for group variable {idx} "
                     f"named {group_name} in expected_groups."
                 )
-            expect_ = _get_expected_groups(b_.data, sort=sort)
+            expect1 = _get_expected_groups(b_.data, sort=sort)
         else:
-            expect_ = expect
-        expect_index = _convert_expected_groups_to_index((expect_,), (isbin_,), sort=sort)[0]
+            expect1 = expect
+        expect_index = _convert_expected_groups_to_index((expect1,), (isbin_,), sort=sort)[0]
```
Comment (dcherian) [inline flox/xarray.py:342]: Prefer `expect_` to `expect1`

**Review by dcherian — COMMENTED** [review]
Looking good. Just some minor comments

File: flox/xarray.py (hunk 7 above)
Comment (Illviljan) [reply flox/xarray.py:342]: I found using numbers were much easier to read when you had multiple `expect` with varying types:
`expect1` similar readabilty to `expect_`
`expect2` much easier to read than `expect__`.

File: flox/core.py (hunk 4 above)
Comment (Illviljan) [reply flox/core.py:105]: I disagree. The point is to indicate that it belongs to the expect-family of types and therefore treated equally.

I find xarray has big issues with this where the same parameter has different typing in other parts of the code. It feels inconsistent and leads to unnecessary bugs in my opinion. It's a reason why we're defining more and more types in  https://github.com/pydata/xarray/blob/main/xarray/core/types.py.

Renaming like this also shows how many different types there are for the expect-family and that maybe a refactor to normalize the types is in order.

File: flox/core.py (hunk 5 above)
Comment (Illviljan) [reply flox/core.py:865]: Probably have to rethink `Aggregation.min_count` then:
https://github.com/xarray-contrib/flox/blob/6cf315a57e006dc01c801039833351240c4431fe/flox/aggregations.py#L210

File: flox/core.py (hunk 6 above)
Comment (Illviljan) [reply flox/core.py]: Nope, doesn't work:
```
flox/core.py:1621: error: Argument 1 to "append" of "list" has incompatible type "Sequence[Any] | ndarray[Any, Any] | Index | None"; expected "Index | None"  [arg-type]
```
Tested in [8e55d3a](https://github.com/xarray-contrib/flox/pull/235/commits/8e55d3a62181153bcd011d99fc73af5b3a01eb6b)

Comment (dcherian) [reply flox/core.py:1622]: Seems like mypy isn't narrowing typing. I'd prefer a `type: ignore` instead.

Comment (Illviljan) [reply flox/core.py]: Add reveal_types:
```python
    out: list[T_ExpectIndexOpt] = []
    for ex, isbin_ in zip(expected_groups, isbin):
        reveal_type(ex) #     flox/core.py:1618: note: Revealed type is "Union[typing.Sequence[Any], numpy.ndarray[Any, Any], pandas.core.indexes.base.Index, None]"
        if isinstance(ex, pd.IntervalIndex) or (isinstance(ex, pd.Index) and not isbin_):
            reveal_type(ex) # flox/core.py:1620: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"
            e = ex
            reveal_type(e) #  flox/core.py:1622: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"
            if sort:
                e = e.sort_values()
            reveal_type(e) #  flox/core.py:1625: note: Revealed type is "Union[pandas.core.indexes.interval.IntervalIndex[Any], pandas.core.indexes.base._IntIndexType]"

            out.append(e)
```

Comment (Illviljan) [reply flox/core.py]: It is the redefinition of ex that's the problem: `ex = ex.sort_values()`.
When redefining a variable mypy always takes the first seen typing of that variable, even though it has already been narrowed down.

At least mypy is consistent with this, it really doesn't like redefinitions with different typing.

I added an `else` without the redefinition instead.

File: flox/core.py (hunk 5 above)
Comment (dcherian) [reply flox/core.py:865]: Can we do `self.min_count : int = 0`, hopefully that's easy to propagate

File: flox/core.py (hunk 6 above)
Comment (dcherian) [reply flox/core.py]: Thanks for tracking that down!

File: flox/core.py (hunk 5 above)
Comment (Illviljan) [reply flox/core.py]: This is as far as I can go with min_count=0. It seems xarray_reduce and group_reduce needs to still be `int | None`.

This is the tricky part:
https://github.com/xarray-contrib/flox/blob/6cf315a57e006dc01c801039833351240c4431fe/flox/core.py#L1930-L1938

Comment (dcherian) [reply flox/core.py]: Ah too bad. Thanks for trying!

File: asv_bench/benchmarks/combine.py [hunk 8]
```diff
@@ -69,7 +70,7 @@ def construct_member(groups):
         ]
         self.kwargs = {
             "agg": flox.aggregations._initialize_aggregation(
-                "sum", "float64", np.float64, 0, None, {}
+                "sum", "float64", np.float64, 0, 0, {}
```
Comment (dcherian) [inline asv_bench/benchmarks/combine.py:73]:
```suggestion
                "sum", "float64", np.float64, 0, None, {}
```

**Pull request merged**